Add support for quantization to `write_weights` and `read_weights`.

adarob opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/tensorflow/tfjs-converter/87)
<!-- Reviewable:end -->

adarob commented:
Works in conjunction with https://github.com/tensorflow/tfjs-core/pull/965.

pyu10055 commented:
Thank you for the contriubiton, a couple high level questions, once the weights are quantized, the weight loader code needs to be able to convert them back to float32 for the op executions? And why the int32 type is not in the quantiziable list?

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[a discussion](https://beta.reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAPE3ip6MSbMF7diAr7:-LAPE3ip6MSbMF7diAr8:b4y9ut5) (no related file):*
and can you provide some description on the quantization algorithm to help me understanding the code bit better? Thanks

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/tensorflow/tfjs-converter/87#-:-LAPCQuwDAREHWsup37z:bmnu2b3)*
<!-- Sent from Reviewable.io -->

adarob commented:
The changes to the loader converts it back to float32 (tensorflow/tfjs-core#965) since this quantization is just to decrease the checkpoint size for now.

I suppose there is no good reason not to support int32, so I've added it :)

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[a discussion](https://beta.reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAPE3ip6MSbMF7diAr7:-LAPzrpv1O4vSUboC9sS:bdprq93) (no related file):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

and can you provide some description on the quantization algorithm to help me understanding the code bit better? Thanks
</blockquote></details>

I added some more documentation. Let me know if it's sufficient.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/tensorflow/tfjs-converter/87#-:-LAPtIs17yYuk3MnRWAu:bz2b9au)*
<!-- Sent from Reviewable.io -->

pyu10055 commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[python/tensorflowjs/write_weights.py, line 139 at r3](https://beta.reviewable.io/reviews/tensorflow/tfjs-converter/87#-LARKgfpC1oBt9akwINf:-LARKgfpC1oBt9akwINg:bie2k5d) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/89be0938c783a3aef9c49619b0ae2bad8582237f/python/tensorflowjs/write_weights.py#L139)):*
> ```Python
>         entry = {
>           'name': 'weight1',
>           'data': np.array([1, 0, 255], 'float32')
> ```

Would it be clear, if keep the data dtype the same as the storage type, and use quantization dtype to indicate the dequant to type?

---

*[python/tensorflowjs/write_weights.py, line 162 at r3](https://beta.reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAROf_g1FFiNcln5W0w:-LAROf_g1FFiNcln5W0x:bq17p3d) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/89be0938c783a3aef9c49619b0ae2bad8582237f/python/tensorflowjs/write_weights.py#L162)):*
> ```Python
>   quantized_group['data'] = quantized_data
>   quantized_group['quantization'] = {
>     'min': m, 'max': M, 'dtype': quantization_dtype.name}
> ```

as mentioned earlier, change to data.dtype.name?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/tensorflow/tfjs-converter/87)*
<!-- Sent from Reviewable.io -->

adarob commented:
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[python/tensorflowjs/write_weights.py, line 139 at r3](https://beta.reviewable.io/reviews/tensorflow/tfjs-converter/87#-LARKgfpC1oBt9akwINf:-LATci1fER3YEbmFmOSI:b-e2jo6h) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/89be0938c783a3aef9c49619b0ae2bad8582237f/python/tensorflowjs/write_weights.py#L139)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

Would it be clear, if keep the data dtype the same as the storage type, and use quantization dtype to indicate the dequant to type?
</blockquote></details>

I agree this would be more clear, but it screws some things up with the typing in the loader code in TS. It's not insurmountable, but I think this is slightly better. Let me know if you disagree.

---

*[python/tensorflowjs/write_weights.py, line 162 at r3](https://beta.reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAROf_g1FFiNcln5W0w:-LATd1buDeUghvjpLIQW:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/89be0938c783a3aef9c49619b0ae2bad8582237f/python/tensorflowjs/write_weights.py#L162)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

as mentioned earlier, change to data.dtype.name?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/tensorflow/tfjs-converter/87)*
<!-- Sent from Reviewable.io -->

adarob commented:
FYI, @nsthorat @pyu10055 this is ready for review.

pyu10055 commented:
Review status: 0 of 7 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[python/tensorflowjs/quantization_util.py, line 54 at r5](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAq2Ynh9txXu4_fA8nQ:-LAq2Ynh9txXu4_fA8nR:bv5qw9h) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/639e794c8b30e41e8f2fade7932894c5eb763067/python/tensorflowjs/quantization_util.py#L54)):*
> ```
> scale
> ```
min_val should be in the original dtype, for same value tensor.

---

*[python/tensorflowjs/quantization_util.py, line 103 at r5](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAq3d0M9oWf_phnvisp:-LAq3d0M9oWf_phnvisq:bie5oqh) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/639e794c8b30e41e8f2fade7932894c5eb763067/python/tensorflowjs/quantization_util.py#L103)):*
> ```Python
>     nudged_zero_point = np.round(quantized_zero_point)
> 
>   # Solve `0 = (0 - nudged_zero_point) * scale + nudged_min`` for `nudged_min`.
> ```

0 = nudged_zero_point * scale + nudged_min?

---

*[python/tensorflowjs/quantization_util_test.py, line 24 at r5](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAq5j_aEHT3InIR1VDe:-LAq5j_aEHT3InIR1VDf:b-frx1t3) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/639e794c8b30e41e8f2fade7932894c5eb763067/python/tensorflowjs/quantization_util_test.py#L24)):*
> ```Python
> class TestQuantizationUtil(unittest.TestCase):
> 
>   def _runQuantizeTest(
> ```

since the dequant code is not the same for ts, it would be better to validate the scale and min as well.

---

*[python/tensorflowjs/read_weights.py, line 27 at r5](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAq7GDR-VBHy6ZsFs9Q:-LAq7GDR-VBHy6ZsFs9R:bv2i1wo) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/639e794c8b30e41e8f2fade7932894c5eb763067/python/tensorflowjs/read_weights.py#L27)):*
> ```Python
> from tensorflowjs import quantization_util
> 
> DTYPE_BYTES = {'float32': 4, 'int32': 4, 'uint8': 2, 'uint16': 4}
> ```

uint8:1, uint16: 2 ?

---

*[python/tensorflowjs/read_weights.py, line 126 at r5](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAq7UwPE1dWfy3WA0Ly:-LAq7UwQEAsoOHnT6Ouf:bk7maj5) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/639e794c8b30e41e8f2fade7932894c5eb763067/python/tensorflowjs/read_weights.py#L126)):*
> ```
> DTYPE_BYTES
> ```
can you use numpy dtype itemsize, instead of manually define the map?

---

*[python/tensorflowjs/read_weights.py, line 138 at r5](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAq8LGB12w4SUUzMUTt:-LAq8LGB12w4SUUzMUTu:bsbcrdf) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/639e794c8b30e41e8f2fade7932894c5eb763067/python/tensorflowjs/read_weights.py#L138)):*
> ```
> weight
> ```
I still think the data dtype needs to be used for reading out the data value, while quant dtype for dequant.

---

*[python/tensorflowjs/write_weights.py, line 139 at r3](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LARKgfpC1oBt9akwINf:-LAq9ydS4l-FrhEeuq_W:b-9f19a8) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/89be0938c783a3aef9c49619b0ae2bad8582237f/python/tensorflowjs/write_weights.py#L139)):*
<details><summary><i>Previously, adarob (Adam Roberts) wrote…</i></summary><blockquote>

I agree this would be more clear, but it screws some things up with the typing in the loader code in TS. It's not insurmountable, but I think this is slightly better. Let me know if you disagree.
</blockquote></details>

looks like you have updated the doc, can you verify if the dtypes are set correctly in the code.

---

*[python/tensorflowjs/write_weights.py, line 27 at r5](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAq9Yox9_KulT8WwHTu:-LAq9Yox9_KulT8WwHTv:bxjjuid) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/639e794c8b30e41e8f2fade7932894c5eb763067/python/tensorflowjs/write_weights.py#L27)):*
> ```
> uint16
> ```
same as about

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-converter/87)*
<!-- Sent from Reviewable.io -->

dsmilkov commented:
Review status: 0 of 7 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[python/tensorflowjs/quantization_util.py, line 28 at r5](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAs2vjY3GK9J1lxNbjR:-LAs2vjY3GK9J1lxNbjS:b-xrr798) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/639e794c8b30e41e8f2fade7932894c5eb763067/python/tensorflowjs/quantization_util.py#L28)):*
> ```Python
>   provided by the `quantization_dtype`.
> 
>   In order to guarantee that 0 is perfectly represented by one of the quanzitzed
> ```

typo: quantized

---

*[python/tensorflowjs/quantization_util.py, line 49 at r5](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAs3ZZE3x6QAFLi5sWK:-LAs3ZZE3x6QAFLi5sWL:b-k2gx12) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/639e794c8b30e41e8f2fade7932894c5eb763067/python/tensorflowjs/quantization_util.py#L49)):*
> ```Python
> 
>   # Compute the min and max for the group.
>   min_val = data.min().astype(np.float64)
> ```

why convert to float64? any problems keeping it the same dtype as `data` (which numpy will do by default for you)?

---

*[python/tensorflowjs/quantization_util.py, line 81 at r5](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAs4fVv9Yst7lCndFfE:-LAs4fVv9Yst7lCndFfF:brl4ot) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/639e794c8b30e41e8f2fade7932894c5eb763067/python/tensorflowjs/quantization_util.py#L81)):*
> ```Python
>   Args:
>     min_val: The actual minimum value of the data.
>     max_val: Thea actual maximum value of the data.
> ```

typo: s/Thea/The

---

*[python/tensorflowjs/quantization_util.py, line 94 at r5](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAs7Xv70Bj8fQOSFkMS:-LAs7Xv70Bj8fQOSFkMT:b-dw2c05) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/639e794c8b30e41e8f2fade7932894c5eb763067/python/tensorflowjs/quantization_util.py#L94)):*
> ```Python
> 
>   quant_max = np.iinfo(quantization_dtype).max
>   scale = (max_val - min_val) / quant_max
> ```

Skip nudging if the values are not crossing zero (e.g. they are all negative , or all positive)

---

*[python/tensorflowjs/quantization_util_test.py, line 33 at r5](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAs6110AKS1plFuvlb_:-LAs6110AKS1plFuvlba:b-byel4d) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/639e794c8b30e41e8f2fade7932894c5eb763067/python/tensorflowjs/quantization_util_test.py#L33)):*
> ```Python
>     de_q = quantization_util.dequantize_weights(q, s, m, data_dtype)
>     if allow_for_min_error:
>       self.assertAlmostEqual(de_q[0], d[0], delta=allow_for_min_error)
> ```

curious as to why the first value gets special treatment?

---

*[python/tensorflowjs/quantization_util_test.py, line 56 at r5](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAs6ilW9r_FBswnm-hx:-LAs6ilW9r_FBswnm-hy:bc3nfee) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/639e794c8b30e41e8f2fade7932894c5eb763067/python/tensorflowjs/quantization_util_test.py#L56)):*
> ```Python
>     # so we have to give the min value a larger error.
>     self._runQuantizeTest(
>         -3, -1, np.float32, np.uint8, allow_for_min_error=1.0)
> ```

This is a suspiciously large error. Back of the envelope math:
You have np.uint8, which gives you 256 buckets, and the range is 2 ((-3) - (-1)), which should give you a max error of `2/256=0.0078125`. Likewise for the other tests. If this is due to the nudging algorithm, there might be a problem in it's implementation. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-converter/87)*
<!-- Sent from Reviewable.io -->

adarob commented:
Review status: 0 of 7 files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[python/tensorflowjs/quantization_util.py, line 28 at r5](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAs2vjY3GK9J1lxNbjR:-LAsFVr63sVwBXUu3ht6:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/639e794c8b30e41e8f2fade7932894c5eb763067/python/tensorflowjs/quantization_util.py#L28)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

typo: quantized
</blockquote></details>

Done.

---

*[python/tensorflowjs/quantization_util.py, line 49 at r5](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAs3ZZE3x6QAFLi5sWK:-LAsFXlm9PLe5nSeHuGg:b-i49cp1) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/639e794c8b30e41e8f2fade7932894c5eb763067/python/tensorflowjs/quantization_util.py#L49)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

why convert to float64? any problems keeping it the same dtype as `data` (which numpy will do by default for you)?
</blockquote></details>

JSON can only handle float64.

---

*[python/tensorflowjs/quantization_util.py, line 54 at r5](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAq2Ynh9txXu4_fA8nQ:-LAsFkxC9F4nkln9kBgM:b-sadqkx) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/639e794c8b30e41e8f2fade7932894c5eb763067/python/tensorflowjs/quantization_util.py#L54)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

> ```
> scale
> ```
min_val should be in the original dtype, for same value tensor.
</blockquote></details>

We need it to be a float64 for JSON, and it's already set above.

---

*[python/tensorflowjs/quantization_util.py, line 81 at r5](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAs4fVv9Yst7lCndFfE:-LAsG-nY9XGgZMtKqYep:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/639e794c8b30e41e8f2fade7932894c5eb763067/python/tensorflowjs/quantization_util.py#L81)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

typo: s/Thea/The
</blockquote></details>

Done.

---

*[python/tensorflowjs/quantization_util.py, line 94 at r5](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAs7Xv70Bj8fQOSFkMS:-LAsG1DJA-DU_sP3tEtd:b-56wrpl) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/639e794c8b30e41e8f2fade7932894c5eb763067/python/tensorflowjs/quantization_util.py#L94)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

Skip nudging if the values are not crossing zero (e.g. they are all negative , or all positive)
</blockquote></details>

I'd prefer that, but it's different from what TF-Lite is doing. Do we want to be consistent?

---

*[python/tensorflowjs/quantization_util.py, line 103 at r5](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAq3d0M9oWf_phnvisp:-LAsGDVi3osdaqAlb-Y1:b-18k6hf) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/639e794c8b30e41e8f2fade7932894c5eb763067/python/tensorflowjs/quantization_util.py#L103)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

0 = nudged_zero_point * scale + nudged_min?
</blockquote></details>

Whoops! No wonder you were confused.

---

*[python/tensorflowjs/quantization_util_test.py, line 24 at r5](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAq5j_aEHT3InIR1VDe:-LAsGV313NWb6zBv5YDt:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/639e794c8b30e41e8f2fade7932894c5eb763067/python/tensorflowjs/quantization_util_test.py#L24)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

since the dequant code is not the same for ts, it would be better to validate the scale and min as well.
</blockquote></details>

Done.

---

*[python/tensorflowjs/quantization_util_test.py, line 33 at r5](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAs6110AKS1plFuvlb_:-LAsGnkQBGTkxNZMDDuW:b-jn5s6n) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/639e794c8b30e41e8f2fade7932894c5eb763067/python/tensorflowjs/quantization_util_test.py#L33)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

curious as to why the first value gets special treatment?
</blockquote></details>

Discussed offline.

---

*[python/tensorflowjs/quantization_util_test.py, line 56 at r5](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAs6ilW9r_FBswnm-hx:-LAsHFYcAr6VstU-04qz:b-jn5s6n) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/639e794c8b30e41e8f2fade7932894c5eb763067/python/tensorflowjs/quantization_util_test.py#L56)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

This is a suspiciously large error. Back of the envelope math:
You have np.uint8, which gives you 256 buckets, and the range is 2 ((-3) - (-1)), which should give you a max error of `2/256=0.0078125`. Likewise for the other tests. If this is due to the nudging algorithm, there might be a problem in it's implementation. 
</blockquote></details>

Discussed offline.

---

*[python/tensorflowjs/read_weights.py, line 27 at r5](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAq7GDR-VBHy6ZsFs9Q:-LAsIVX5E-lOHLi0QOZ4:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/639e794c8b30e41e8f2fade7932894c5eb763067/python/tensorflowjs/read_weights.py#L27)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

uint8:1, uint16: 2 ?
</blockquote></details>

Done.

---

*[python/tensorflowjs/read_weights.py, line 126 at r5](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAq7UwPE1dWfy3WA0Ly:-LAsIW4g8QmdWPq43zM5:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/639e794c8b30e41e8f2fade7932894c5eb763067/python/tensorflowjs/read_weights.py#L126)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

> ```
> DTYPE_BYTES
> ```
can you use numpy dtype itemsize, instead of manually define the map?
</blockquote></details>

Done.

---

*[python/tensorflowjs/read_weights.py, line 138 at r5](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAq8LGB12w4SUUzMUTt:-LAsIhmx3ql_0FAFS-bU:b-wpz2u4) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/639e794c8b30e41e8f2fade7932894c5eb763067/python/tensorflowjs/read_weights.py#L138)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

> ```
> weight
> ```
I still think the data dtype needs to be used for reading out the data value, while quant dtype for dequant.
</blockquote></details>

I'm not quite sure what you mean. We need to pass in the final dtype here for it to convert to it.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-converter/87)*
<!-- Sent from Reviewable.io -->

pyu10055 commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> Thanks you!

---

Review status: 0 of 7 files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[python/tensorflowjs/quantization_util.py, line 101 at r6](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAsdjzx60zyfjj98CRw:-LAsdjzyBOJnOAHor1oF:bs51s9q) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/6c0c4004e481d6cf14fb16689a876f994650704d/python/tensorflowjs/quantization_util.py#L101)):*
> ```Python
>   if min_val <= 0 <= max_val:
>     quantized_zero_point = (0 - min_val) / scale
>     if quantized_zero_point < 0:
> ```

when 0 is in the range, the quantized_zero_point should allow be with the quant range.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-:-LAsdxLkBaJOHhRM0dJ-:b-jlogt6)*
<!-- Sent from Reviewable.io -->

pyu10055 commented:
Reviewed 2 of 7 files at r4, 1 of 5 files at r5, 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-converter/87)*
<!-- Sent from Reviewable.io -->

adarob commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[python/tensorflowjs/quantization_util.py, line 101 at r6](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAsdjzx60zyfjj98CRw:-LAsfnMc3ooPNUbzMzQz:bdcv6kv) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/6c0c4004e481d6cf14fb16689a876f994650704d/python/tensorflowjs/quantization_util.py#L101)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

when 0 is in the range, the quantized_zero_point should allow be with the quant range.
</blockquote></details>

Good point!

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-converter/87)*
<!-- Sent from Reviewable.io -->

caisq commented:
Thanks for starting this effort. I think quantization is very important and we should definitely support it. But as one of my comments below states, there are implications of this for not only Python, but also JavaScript side of TensorFlow.js. Please let us know your plan to support (or not support) saving/loading quantized weights from/to JavaScript. The complexity of this feels high enough to merit some sort of design process.

---

Review status: 6 of 7 files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[python/tensorflowjs/quantization_util.py, line 1 at r6](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAsPlyD8hhO8gRBnPYt:-LAsPlyE52D6sSBEXS7M:b5x4vwf) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/6c0c4004e481d6cf14fb16689a876f994650704d/python/tensorflowjs/quantization_util.py#L1)):*
> ```
> # Copyright 2018 Google LLC
> ```

Comment on file name: how about just call this file "quantization.py" (and the BUILD target just "quantization") for simplicitiy?

---

*[python/tensorflowjs/quantization_util.py, line 18 at r6](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAsUAQb5ww_Tb7O5wDX:-LAsUAQb5ww_Tb7O5wDY:b-pdrbzz) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/6c0c4004e481d6cf14fb16689a876f994650704d/python/tensorflowjs/quantization_util.py#L18)):*
> ```
> QUANTIZATION_DTYPES
> ```

nit: if not used outside this module, this should be named `_QUANTIZATION_DTYPES` for privacy.

---

*[python/tensorflowjs/quantization_util.py, line 35 at r6](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAsg8hg3jac-EFTdcgN:-LAsg8hg3jac-EFTdcgO:b-83gks6) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/6c0c4004e481d6cf14fb16689a876f994650704d/python/tensorflowjs/quantization_util.py#L35)):*
> ```
> weights
> ```

Wrong arg name.

---

*[python/tensorflowjs/quantization_util.py, line 92 at r6](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAsgoaoBJl8hCJ1U0Mf:-LAsgoaoBJl8hCJ1U0Mg:baly725) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/6c0c4004e481d6cf14fb16689a876f994650704d/python/tensorflowjs/quantization_util.py#L92)):*
> ```
>  """
> ```

nit: Add `Raises` section to doc string.

---

*[python/tensorflowjs/quantization_util_test.py, line 35 at r6](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAshST-8mSPnSmGO4t6:-LAshST-8mSPnSmGO4t7:b-4f7yjx) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/6c0c4004e481d6cf14fb16689a876f994650704d/python/tensorflowjs/quantization_util_test.py#L35)):*
> ```
> d_0 = np.zeros(1, data_dtype)
> ```

Is there an assertion that the nudging effectively preserves the zeros? If not, add it?

---

*[python/tensorflowjs/read_weights.py, line 120 at r6](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAshvbHDLKOwsSCICSm:-LAshvbHDLKOwsSCICSn:bircmjw) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/6c0c4004e481d6cf14fb16689a876f994650704d/python/tensorflowjs/read_weights.py#L120)):*
> ```
> quantization = weight.get('quantization', None)
> ```

So this changes the weight manifest format. As the manifest format is used in both Python and JavaScript, the code and documentation in JavaScript/TypeScript needs to be updated as well. Do you plan to support loading / saving quantized weights to / from JavaScript? The complexity might merrit a design doc.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-:-LAsik9_4kEosx5JUNsa:b-n6olvo)*
<!-- Sent from Reviewable.io -->

dsmilkov commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> Nice work Adam! Couple minor comments - wait for Shanqing's LGTM also.

---

Reviewed 2 of 9 files at r4, 6 of 10 files at r7.
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[python/tensorflowjs/write_weights.py, line 114 at r7](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAt89VHCWmW_hCmPVsW:-LAt89VI004XW64O1ejz:b-slw2z6) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/100e7f87d0b5ecd26ce2a71ee2eb23faeb558341/python/tensorflowjs/write_weights.py#L114)):*
> ```Python
>   for group_index, group in enumerate(weight_groups):
>     if quantization_dtype:
>       group = _quantize_group(group, quantization_dtype)
> ```

Minor comment: this method feels like the quantization params (min and scale) are across the whole group, instead of per-entry. Perhaps rename _quantize_group to _quantize_entry and do the for loop over all groups here?

---

*[python/tensorflowjs/quantization_util_test.py, line 35 at r6](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAshST-8mSPnSmGO4t6:-LAt3-yiBTSdBoB50uXb:bwr9rad) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/6c0c4004e481d6cf14fb16689a876f994650704d/python/tensorflowjs/quantization_util_test.py#L35)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

> ```
> d_0 = np.zeros(1, data_dtype)
> ```

Is there an assertion that the nudging effectively preserves the zeros? If not, add it?
</blockquote></details>

I think the assert is right under this comment. A related q: Since It doesn't make sense to assert preserving of zeros when zero is not in the natural range, should we check for that here?

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-:-LAt8R1141_4VLLLEEaT:b-m0wfvp)*
<!-- Sent from Reviewable.io -->

dsmilkov commented:
Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[python/tensorflowjs/quantization_util.py, line 94 at r5](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAs7Xv70Bj8fQOSFkMS:-LAt8e6n7HbEhRIAt07s:byrrnzu) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/639e794c8b30e41e8f2fade7932894c5eb763067/python/tensorflowjs/quantization_util.py#L94)):*
<details><summary><i>Previously, adarob (Adam Roberts) wrote…</i></summary><blockquote>

I'd prefer that, but it's different from what TF-Lite is doing. Do we want to be consistent?
</blockquote></details>

Discussed offline. Resolving

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-converter/87)*
<!-- Sent from Reviewable.io -->

dsmilkov commented:
Reviewed 2 of 3 files at r8.
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[python/tensorflowjs/quantization_test.py, line 29 at r8](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAxpYgd-wYceCSaI7tc:-LAxpYge95yrJK1fm4tE:bkwckd9) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/a4e911204981d58eaa25fd8e953e5ddc59642dda/python/tensorflowjs/quantization_test.py#L29)):*
> ```Python
>     d = np.arange(range_min, range_max + 1, dtype=data_dtype)
>     q, s, m = quantization.quantize_weights(d, quantization_dtype)
>     self.assertAlmostEqual(s, expected_scale)
> ```

Travis fails here for python2.7. I think it's because division works differently between Py2 and 3. Add
```
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
```

to all python files

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-converter/87)*
<!-- Sent from Reviewable.io -->

adarob commented:
Review status: 5 of 7 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[python/tensorflowjs/quantization_test.py, line 29 at r8](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAxpYgd-wYceCSaI7tc:-LAxz9EV8qLzRh1tbOVD:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/a4e911204981d58eaa25fd8e953e5ddc59642dda/python/tensorflowjs/quantization_test.py#L29)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

Travis fails here for python2.7. I think it's because division works differently between Py2 and 3. Add
```
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
```

to all python files
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-converter/87)*
<!-- Sent from Reviewable.io -->

caisq commented:
Review status: 5 of 7 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[a discussion](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAPE3ip6MSbMF7diAr7:-LAy-T2WAmIVw_w82QYa:blhktvq) (no related file):*
<details><summary><i>Previously, adarob (Adam Roberts) wrote…</i></summary><blockquote>

I added some more documentation. Let me know if it's sufficient.
</blockquote></details>

Another comment: @adarob I noticed that the support for quantization is not yet hooked up with the tensorflowjs_converter binary yet. As far as I'm concerned, you can either do it in this PR or as another PR. 

See:
https://github.com/tensorflow/tfjs-converter/blob/master/python/tensorflowjs/converters/converter.py#L75

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-converter/87)*
<!-- Sent from Reviewable.io -->

dsmilkov commented:
Review status: 5 of 7 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[python/tensorflowjs/write_weights.py, line 20 at r9](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAyXQgcBFar4iE5CIz0:-LAyXQgcBFar4iE5CIz1:bqqeod2) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/168717b0d28ec314a8e77a9a67fc43bea5ffdd6a/python/tensorflowjs/write_weights.py#L20)):*
> ```Python
> import math
> import os
> ```

can you add the 3 future imports here (seems like we forgot to add them originally) and check for the other files as well. Thanks!

---

*[python/tensorflowjs/write_weights.py, line 22 at r9](https://reviewable.io/reviews/tensorflow/tfjs-converter/87#-LAyXeRC-G4vNMBXQWS4:-LAyXeRC-G4vNMBXQWS5:b-p0iqbp) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/168717b0d28ec314a8e77a9a67fc43bea5ffdd6a/python/tensorflowjs/write_weights.py#L22)):*
> ```Python
> 
> import numpy as np
> from tensorflowjs import quantization
> ```

this fails on py2.7. Credits go to @caisq: you should add
`from tensorflowjs import quantization`
in the `__init__` file

https://github.com/tensorflow/tfjs-converter/blob/master/python/tensorflowjs/__init__.py#L21

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-converter/87)*
<!-- Sent from Reviewable.io -->
